File: RestaurantManager/Controllers/RestaurantController.cs
using Restaurant.Core;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;

namespace RestaurantManager.Controllers
{
    public class RestaurantController : Controller
    {
        private RestaurantMenuContext _dbContext;

        public RestaurantController()
        {
            _dbContext = new RestaurantMenuContext();
        }
        
        public ActionResult Index()
        {
            return View();
        }

        public JsonResult GetRestaurants()
        {
            return Json(_dbContext.Restaurants.ToList(), JsonRequestBehavior.AllowGet);
        }

        public JsonResult GetRestaurantById(int id)
        {
            return Json(_dbContext.Restaurants.FirstOrDefault(r => r.Id == id), JsonRequestBehavior.AllowGet);
        }

        public JsonResult GetRestaurantByName(string name)
        {
            return Json(_dbContext.Restaurants.FirstOrDefault(r => r.Name == name), JsonRequestBehavior.AllowGet);
        }

        public JsonResult GetRestaurantsByName(string name)
        {
            return Json(_dbContext.Restaurants.Where(r => r.Name.Contains(name)), JsonRequestBehavior.AllowGet);
        }

        public JsonResult AddRestaurant(Restaurant.Core.Restaurant restaurant)
        {
            _dbContext.Restaurants.Add(restaurant);
            _dbContext.SaveChanges();

            return GetRestaurants();
        }

        public JsonResult UpdateRestaurant(Restaurant.Core.Restaurant restaurant)
        {
            _dbContext.Entry(restaurant).State = System.Data.Entity.EntityState.Modified;
            _dbContext.SaveChanges();

            return GetRestaurants();
        }

        public JsonResult DeleteRestaurant(int id)
        {
            var restaurant = _dbContext.Restaurants.FirstOrDefault(r => r.Id == id);

            if (restaurant != null)
            {
                foreach (var dish in _dbContext.Dishes.Where(f => f.RestaurantId == id))
                {
                    _dbContext.Dishes.Remove(dish);
                }

                _dbContext.Restaurants.Remove(restaurant);
            }

            _dbContext.SaveChanges();

            return GetRestaurants();
        }
    }
}

File: RestaurantManager/app/controllers/restaurant.controller.js
(function () {
    'use strict';

    angular
        .module('app')
        .controller('RestaurantController', RestaurantController);

    function RestaurantController($scope, $location, $routeParams, RestaurantDataService) {
        
        $scope.restaurants = getRestaurants();

        if ($routeParams && $routeParams.id)
            getRestaurantById();

        $scope.restaurantService = {
            getRestaurants: getRestaurants,
            getRestaurantById: getRestaurantById,
            getRestaurantByName: getRestaurantByName,
            addRestaurant: addRestaurant,
            updateRestaurant: updateRestaurant,
            deleteRestaurant: deleteRestaurant,
            filterRestaurantsByName: filterRestaurantsByName
        }

        $scope.goAddRestaurant = function () {
            $location.path('/AddRestaurant/');
        };

        $scope.goRestaurantList = function () {
            $location.path('/RestaurantList/');
        }

        function addRestaurant() {
            RestaurantDataService
                 .addRestaurant($scope.restaurant)
                 .then(function (restaurants) {
                     $location.path('/RestaurantList');
                     alert('Novo restaurante registrado com sucesso.');                     
                     $scope.restaurants = restaurants;
                     $scope.restaurant = {};
                 });
        }

        function deleteRestaurant(id) {
            RestaurantDataService
                .deleteRestaurant(id)
                .then(function (restaurants) {
                    $location.path('/RestaurantList');
                    alert('Restaurant com Id: ' + id + ' deletado com sucesso');                    
                    $scope.restaurants = restaurants;
                });
        }

        function updateRestaurant() {
            var req = {
                Id: $scope.restaurant.Id,
                Name: $scope.restaurant.Name
            };

            RestaurantDataService
                .updateRestaurant(req)
                .then(function (restaurants) {                    
                    $location.path('/RestaurantList');
                    alert('Alteração realizada com sucesso.');
                    $scope.restaurants = restaurants;
                });
        }

        function getRestaurants() {
            RestaurantDataService
                .getRestaurants()
                .then(function (restaurants) {
                    $scope.restaurants = restaurants;
                });
        }

        function getRestaurantById(id) {
            RestaurantDataService
                .getRestaurantById(id || $routeParams.id)
                .then(function (restaurant) {
                    $scope.restaurant = restaurant;
                });
        }

        function getRestaurantByName(name) {
            RestaurantDataService
                .getRestaurantByName(name || $routeParams.name)
                .then(function (restaurant) {
                    $scope.restaurant = restaurant;
                });
        }

        function filterRestaurantsByName(name) {
            if (name && name.length)
                RestaurantDataService
                    .getRestaurantsByName(name || $routeParams.name)
                    .then(function (restaurants) {
                        $scope.restaurants = restaurants;
                    });
            else
                getRestaurants();
        }
    }
})();
